src/main/java/dan200/computercraft/shared/peripheral/speaker/SpeakerPeripheral.java

```diff
@@ -0,0 +1,264 @@
+    {
+    }
+
+    @Override
+    public void detach(@Nonnull IComputerAccess computerAccess)
+    {
+    }
+
+    @Nonnull
+    @Override
+    public String getType()
+    {
+        return "speaker";
+    }
+
+    @Nonnull
+    @Override
+    public String[] getMethodNames()
+    {
+        return new String[] {
+                "playSound", // Plays sound at resourceLocator
+                "playNote" // Plays note
+        };
+    }
+
+    @Override
+    public Object[] callMethod(@Nonnull IComputerAccess computerAccess, @Nonnull ILuaContext context, int methodIndex, @Nonnull Object[] args) throws LuaException
+    {
+        switch (methodIndex)
+        {
+            // playsound
+            case 0:
+            {
+                return playSound(args, context, false);
+            }
+
+            // playnote
+            case 1:
+            {
+                return playNote(args, context);
+            }
+
+            default:
+            {
+                throw new LuaException("Method index out of range!");
+            }
+
+        }
+    }
+
+    @Nonnull
+    private synchronized Object[] playNote(Object[] arguments, ILuaContext context) throws LuaException
+    {
+        double volume = 1f;
+        double pitch = 1f;
+
+        // Check if arguments are correct
+        if (arguments.length == 0) // Too few args
+        {
+            throw new LuaException("Expected string, number (optional), number (optional)");
+        }
+
+        if (!(arguments[0] instanceof String)) // Arg wrong type
+        {
+            throw new LuaException("Expected string, number (optional), number (optional)");
+        }
+
+        if (!SoundEvent.REGISTRY.containsKey(new ResourceLocation("block.note." + arguments[0])))
+        {
+            throw new LuaException("Invalid instrument, \"" + arguments[0] + "\"!");
+        }
+
+        if (arguments.length > 1)
+        {
+            if (!(arguments[1] instanceof Double) && arguments[1] != null)   // Arg wrong type
+            {
+                throw new LuaException("Expected string, number (optional), number (optional)");
+            }
+            volume = arguments[1] != null ? ((Double) arguments[1]).floatValue() : 1f;
+
+        }
+
+        if (arguments.length > 2)
+        {
+            if (!(arguments[1] instanceof Double) && arguments[2] != null)  // Arg wrong type
+            {
+                throw new LuaException("Expected string, number (optional), number (optional)");
+            }
+            pitch = arguments[2] != null ? ((Double) arguments[2]).floatValue() : 1f;
+        }
+
+        // If the resource location for note block notes changes, this method call will need to be updated
+        Object[] returnValue = playSound(new Object[] {"block.note." + arguments[0], volume, Math.pow(2d, (pitch - 12) / 12d)}, context, true);
+
+        if (returnValue[0] instanceof Boolean && (Boolean) returnValue[0])
+        {
+            m_notesThisTick++;
```

Comment:
This seems to only be incremented in playNote, not playSound. Surely this means playSound() isn't rate limited?